src/EventStore.Core/DataStructures/ProbabilisticFilter/MemoryMappedFileBloomFilter/MemoryMappedFileBloomFilter.cs

```diff
@@ -0,0 +1,279 @@
+			long hash2 = ((long)_hashers[2].Hash(bytes) << 32) | _hashers[3].Hash(bytes);
+
+			// guarantee our LOADs are executed now and not prior.
+			// dual with the memory barrier implicit in the Add(...) method
+			Thread.MemoryBarrier();
+
+			long hash = hash1;
+			for (int i = 0; i < NumHashFunctions; i++) {
+				hash += hash2;
+				hash &= long.MaxValue; // make non-negative
+				long bitPosition = hash % _numBits;
+				if (!IsBitSet(bitPosition))
+					return false;
+			}
+
+			return true;
+		}
+
+		private static long GetBytePositionInFile(long bitPosition) {
+			var bytePositionInFile = bitPosition / 8;
+
+			// account for 4 bytes at the end of each cache line being a hash
+			bytePositionInFile += bytePositionInFile / (CacheLineSize - BloomFilterIntegrity.HashSize) * BloomFilterIntegrity.HashSize;
+
+			// first cache line is reserved for the header
+			bytePositionInFile += CacheLineSize;
+
+			return bytePositionInFile;
+		}
+
+		private ref byte ReadByte(long bytePositionInFile) {
+			EnsureBounds(bytePositionInFile, length: 1);
+			return ref *(_pointer + bytePositionInFile);
+		}
+
+		private Span<byte> ReadCacheLineFor(long bytePositionInFile) {
+			bytePositionInFile = bytePositionInFile / CacheLineSize * CacheLineSize;
+			EnsureBounds(bytePositionInFile, CacheLineSize);
+			return new Span<byte>(_pointer + bytePositionInFile, CacheLineSize);
+		}
+
+		private void EnsureBounds(long bytePosition, int length) {
+			if (bytePosition > _fileSize - length)
+				throw new ArgumentOutOfRangeException(
+					nameof(bytePosition),
+					bytePosition,
+					$"{bytePosition:N0}/{_fileSize:N0}");
+		}
+
+		public void Flush() {
+			// need to be sure the read operations were not done in advance of this point
+			// not 100% sure, however, that the barrier is required.
+			Thread.MemoryBarrier();
+			_mmfWriteAccessor.Flush();
+			_fileStream.FlushToDisk();
+		}
+
+		public void Dispose() {
+			if (_disposed)
+				return;
+
+			_disposed = true;
+
+			_mmfWriteAccessor.SafeMemoryMappedViewHandle.ReleasePointer();
+			_mmfWriteAccessor?.Dispose();
+			_mmf?.Dispose();
+		}
+
+		private void FillWithZeros() {
+			Log.Debug("Zeroing bloom filter...");
+			var bytesToClear = _fileSize;
+			var startAddress = _pointer;
+			while (bytesToClear > 0) {
+				uint bytesToClearInBlock = bytesToClear > uint.MaxValue ? uint.MaxValue : (uint)bytesToClear;
+				Unsafe.InitBlock(ref *startAddress, 0, bytesToClearInBlock);
+				startAddress += bytesToClearInBlock;
+				bytesToClear -= bytesToClearInBlock;
+			}
+			Log.Debug("Done zeroing bloom filter");
+		}
+
+		// corruptionThreshold = 0.05 => tolerate up to 5% corruption
+		public void Verify(double corruptionThreshold) {
+			Ensure.Nonnegative(corruptionThreshold, nameof(corruptionThreshold));
+
+			Log.Debug("Verifying bloom filter...");
+			var corruptedCacheLines = 0;
+
+			for (int i = CacheLineSize; i < _fileSize; i += CacheLineSize) {
```

Comment:
```suggestion
			for (long i = CacheLineSize; i < _fileSize; i += CacheLineSize) {
```

`i` may go out of bounds for large bloom filters

Reply:
👍 good catch in case we ever allow a bloom filter that large

Reply:
@timothycoleman the max size is currently set to 4GB which will overflow an int